lang/java/avro/src/main/java/org/apache/avro/generic/GenericData.java

```diff
@@ -94,24 +95,38 @@ public GenericData(ClassLoader classLoader) {
   /** Return the class loader that's used (by subclasses). */
   public ClassLoader getClassLoader() { return classLoader; }
 
-  public Map<String, Conversion<?>> conversions =
+  private Map<String, Conversion<?>> conversions =
       new HashMap<String, Conversion<?>>();
 
-  public Map<Class<?>, Conversion<?>> conversionsByClass =
-      new IdentityHashMap<Class<?>, Conversion<?>>();
+  private Map<Class<?>, List<Conversion<?>>> conversionsByClass =
+      new IdentityHashMap<Class<?>, List<Conversion<?>>>();
+
+  public Iterable<Conversion<?>> getAllConversions() {
+    return conversions.values();
+  }
 
   public void addLogicalTypeConversion(Conversion<?> conversion) {
     conversions.put(conversion.getLogicalTypeName(), conversion);
-    conversionsByClass.put(conversion.getConvertedType(), conversion);
+    Class<?> type = conversion.getConvertedType();
+    if (conversionsByClass.containsKey(type)) {
```

Comment:
I don't think it needs to be. Users generally won't be changing the behavior of the singleton class unless for some reason they want conversions everywhere the singleton is used instead of passing a GenericData instance around. There is no method to remove conversions, so this is something that should be done during setup before anything else is using the singleton.